### PR TITLE
dynamic object related fixes

### PR DESCRIPTION
**Pull request by npitre**
- userspace: plug thread index leak in k_object_alloc()
- tests: dynamic_thread: fix test_thread_index_management
- tests: dynamic_thread: no need to exclude x86 anymore

### CODE REVIEW

**Review by carlocaione — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

File: tests/kernel/threads/dynamic_thread/prj.conf [hunk 1]
```diff
@@ -1,3 +1,3 @@
 CONFIG_ZTEST=y
 CONFIG_TEST_USERSPACE=y
-CONFIG_HEAP_MEM_POOL_SIZE=8192
+CONFIG_HEAP_MEM_POOL_SIZE=20000
```
Comment (andyross) [inline tests/kernel/threads/dynamic_thread/prj.conf:3]: Must it be this big?  I guess all our userspace platforms are mostly unconstrained, but there are still small systems in the world that might want to run this test.

File: tests/kernel/threads/dynamic_thread/testcase.yaml [hunk 2]
```diff
@@ -2,4 +2,3 @@ tests:
   kernel.threads.dynamic:
     tags: kernel threads userspace ignore_faults
     filter: CONFIG_ARCH_HAS_USERSPACE
-    arch_exclude: x86 # see GitHub #17893
```
Comment (andyross) [inline tests/kernel/threads/dynamic_thread/testcase.yaml:5]: Did we ever update the dynamic allocator to use the aligned_alloc API though?

**Review by andyross — APPROVED** [review]
A few questions were prompted, but looks correct.

**Comment by npitre** [comment]
Of course I should have written "32KB" above, not "32MB".

**Pull request merged**